### PR TITLE
Add changelog for 1.12

### PR DESCRIPTION
**Pull request by brson**
[Rendered](https://github.com/brson/rust/blob/relnotes/RELEASES.md).

### CODE REVIEW

File: RELEASES.md [hunk 1]
```diff
@@ -1,3 +1,263 @@
...
+  (https://github.com/rust-lang/rust/pull/34946)
+* [`btree_map::Entry`, `btree_map::VacantEntry` and `btree_map::OccupiedEntry`
+  implement `Debug`]
+  (https://github.com/rust-lang/rust/pull/34885)
+* [`String` implements `AddAssign`]
+  (https://github.com/rust-lang/rust/pull/34890)
+* [Variadic `extern fn` pointers implement the `Clone`, `PartialEq`, `Eq`,
+  `PartialOrd`, `Ord`, `Hash`, `fmt::Pointer`, and `fmt::Debug` traits]
+  (https://github.com/rust-lang/rust/pull/34879)
+* [`FileType implements `Debug`]
```
Comment (retep998) [inline RELEASES.md]: Missing backtick

**Review by retep998 — CHANGES_REQUESTED** [review]

File: RELEASES.md [hunk 2]
```diff
@@ -1,3 +1,263 @@
...
+* [Rust releases now come with source packages that can be installed by rustup
+  via `rustup component add rust-src`]
+  (https://github.com/rust-lang/rust/pull/34366).
+  The resulting source code can be used by tools and IDES, located in the
+  sysroot under `lib/rustlib/src`.
+
+Misc
+----
+
+* [The compiler can new be built against LLVM 3.9]
```
Comment (WiSaGaN) [inline RELEASES.md]: s/new/now

File: RELEASES.md [hunk 3]
```diff
@@ -1,3 +1,263 @@
...
+* [In error type descriptions, unknown numeric types are named `{integer}` or
+  `{float}` instead of `_`]
+  (https://github.com/rust-lang/rust/pull/35080)
+* [`rustc` emits a clearer error when inner attributes follow a doc comment]
+  (https://github.com/rust-lang/rust/pull/34676)
+
+Language
+--------
+
+* [`macro_rules!` invocations can be made within `macro_rules! invocations]
```
Comment (killercup) [inline RELEASES.md]: Missing backtick after second macro_rules

File: RELEASES.md [hunk 4]
```diff
@@ -1,3 +1,263 @@
...
+  producing inconsistent results]
+  (https://github.com/rust-lang/rust/pull/35267)
+* [The `mips-unknown-linux-gnu` target uses hardware floating point by default]
+  (https://github.com/rust-lang/rust/pull/34910)
+* [The `rustc` arguments, `--print target-cpus`, `--print target-features`,
+  `--print relocation-models`, and `--print code-models` print the available
+  options to the `-C target-cpu`, `-C target-feature`, `-C relocation-model` and
+  `-C code-model` code generation arguments]
+  (https://github.com/rust-lang/rust/pull/34845)
+* [`rustc` supports three new MUSL targets on ARM: `arm-unknown-linux-musleabi`,
```
Comment (japaric) [inline RELEASES.md]: perhaps should mention that these targets produce *statically* linked binaries.

**Comment by brson** [comment]
Updated. Thanks for the feedback.

File: RELEASES.md [hunk 5]
```diff
@@ -1,3 +1,264 @@
...
+  in multiple styles]
+  (https://github.com/rust-lang/rust/pull/33642)
+* [The lifetime bounds on `[T]::binary_search_by` and
+  `[T]::binary_search_by_key` have been adjusted to be more flexible]
+  (https://github.com/rust-lang/rust/pull/34762)
+* [`Option` implements `From` for its contained type]
+  (https://github.com/rust-lang/rust/pull/34828)
+* [`Cell`, `RefCell` and `UnsafeCell` implement `From` for their contained type]
+  (https://github.com/rust-lang/rust/pull/35392)
+* [`RwLock` panics of the reader count overflows]
```
Comment (sfackler) [inline RELEASES.md]: "if"

File: RELEASES.md [hunk 6]
```diff
@@ -1,3 +1,264 @@
...
+  (https://github.com/rust-lang/cargo/pull/2867)
+* [Add `--dry-run` to `cargo publish`]
+  (https://github.com/rust-lang/cargo/pull/2849)
+* [Add support for `RUSTDOCFLAGS`]
+  (https://github.com/rust-lang/cargo/pull/2794)
+
+Performance
+-----------
+
+* [`rustc` produces more compact code by more precisely identifying the live
```
Comment (arielb1) [inline RELEASES.md]: That's only for MIR, and no improvement over pre-MIR.

**Review by arielb1 — CHANGES_REQUESTED** [review]

File: RELEASES.md [hunk 7]
```diff
@@ -1,3 +1,264 @@
...
+--------
+
+* [`rustc` translates code to LLVM IR via its own "middle" IR (MIR)]
+  (https://github.com/rust-lang/rust/pull/34096).
+  This translation pass is far simpler than the previous AST->LLVM pass, and
+  creates opportunities to perform new optimizations directly on the MIR. It
+  was previously described [on the Rust blog]
+  (https://blog.rust-lang.org/2016/04/19/MIR.html).
+* [Print the Rust target name, not the LLVM target name, with
+  `--print-target-list`]
```
Comment (liigo) [inline RELEASES.md]: `--print    target-list`

File: RELEASES.md [hunk 8]
```diff
@@ -1,3 +1,262 @@
...
+
+* [`macro_rules!` invocations can be made within `macro_rules!` invocations]
+  (https://github.com/rust-lang/rust/pull/34925)
+* [`macro_rules!` meta-variables are hygienic]
+  (https://github.com/rust-lang/rust/pull/35453)
+* [`macro_rules!` `tt` matchers can be reparsed correctly, making them much more
+  useful]
+  (https://github.com/rust-lang/rust/pull/34908)
+* [`macro_rules!` `stmt` matchers correctly consume the entire contents when
+  insider non-braces invocations]
```
Comment (wesleywiser) [inline RELEASES.md]: I think this should be "inside" instead of "insider"

**Comment by brson** [comment]
Updated.

**Comment by brson** [comment]
Now this needs to be backported to the _1.13_ beta as well.

**Pull request merged**